extensions/pkg/controller/worker/genericactuator/actuator.go

```diff
@@ -252,6 +255,30 @@ func (a *genericActuator) shallowDeleteMachineClassSecrets(ctx context.Context,
 	return nil
 }
 
+// removeFinalizerFromWorkerSecretRef removes the MCM finalizers from the secret that is referenced by the worker
+func (a *genericActuator) removeFinalizerFromWorkerSecretRef(ctx context.Context, logger logr.Logger, worker *extensionsv1alpha1.Worker) error {
+	logger.Info("Removing MCM finalizers from worker`s secret")
+	secret, err := kutil.GetSecretByReference(ctx, a.client, &worker.Spec.SecretRef)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	if controllerutil.ContainsFinalizer(secret, mcmFinalizer) {
+		if err := controllerutils.PatchRemoveFinalizers(ctx, a.client, secret, mcmFinalizer); err != nil {
+			return err
+		}
+	}
+	if controllerutil.ContainsFinalizer(secret, mcmProviderFinalizer) {
+		if err := controllerutils.PatchRemoveFinalizers(ctx, a.client, secret, mcmProviderFinalizer); err != nil {
+			return err
+		}
+	}
+
+	return nil
```

Comment:
Some improvement suggestions
```suggestion
	finalizersToRemove := []string{}
	if controllerutil.ContainsFinalizer(secret, mcmFinalizer) {
		finalizersToRemove = append(finalizersToRemove, mcmFinalizer)
	}
	if controllerutil.ContainsFinalizer(secret, mcmProviderFinalizer) {
		finalizersToRemove = append(finalizersToRemove, mcmProviderFinalizer)
	}
	if len(finalizersToRemove) == 0 {
		return nil
	}

	return controllerutils.PatchRemoveFinalizers(ctx, a.client, secret, finalizersToRemove...)
```

Some improvement suggestions.